etna/models/nn/patchts.py

```diff
@@ -0,0 +1,379 @@
+from etna.distributions import IntDistribution
+from etna.models.base import DeepBaseModel
+from etna.models.base import DeepBaseNet
+
+if SETTINGS.torch_required:
+    import torch
+    import torch.nn as nn
+
+
+class PatchTSBatch(TypedDict):
+    """Batch specification for PatchTS."""
+
+    encoder_real: "torch.Tensor"
+    decoder_real: "torch.Tensor"
+    encoder_target: "torch.Tensor"
+    decoder_target: "torch.Tensor"
+    segment: "torch.Tensor"
+
+
+class PositionalEncoding(nn.Module):
+
+    def __init__(self, d_model: int, dropout: float = 0.1, max_len: int = 5000):
+        super().__init__()
+        self.dropout = nn.Dropout(p=dropout)
+
+        position = torch.arange(max_len).unsqueeze(1)
+        div_term = torch.exp(torch.arange(0, d_model, 2) * (-math.log(10000.0) / d_model))
+        pe = torch.zeros(max_len, 1, d_model)
+        pe[:, 0, 0::2] = torch.sin(position * div_term)
+        pe[:, 0, 1::2] = torch.cos(position * div_term)
+        self.register_buffer('pe', pe)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Arguments:
+            x: Tensor, shape ``[batch_size, input_size, patch_num, embedding_dim]``
+        """
+        x = torch.reshape(x, (x.shape[0] * x.shape[1], x.shape[2], x.shape[3]))
+        # x.shape = (batch_size * input_size, patch_num, embedding_dim)
+        x = x.permute(1, 0, 2)  # (patch_num, batch_size * input_size, embedding_dim)
+        x = x + self.pe[:x.size(0)]
+        return self.dropout(x)
+
+
+class PatchTSNet(DeepBaseNet):
+    """PatchTS based Lightning module"""
+
+    def __init__(
+            self,
+            encoder_length: int,
+            patch_len: int,
+            stride: int,
+            num_layers: int,
+            hidden_size: int,
+            feedforward_size: int,
+            nhead: int,
+            lr: float,
+            loss: "torch.nn.Module",
+            optimizer_params: Optional[dict],
+    ) -> None:
+        """Init PatchTS
+
+        Parameters
+        ----------
+        encoder_length:
+            encoder length
+        patch_len:
+            size of patch
+        stride:
+            step of patch
+        num_layers:
+            number of layers
+        hidden_size:
+            size of the hidden state
+        feedforward_size:
+            size of feedforward layers in transformer
+        nhead:
+            number of transformer heads
+        lr:
+            learning rate
+        loss:
+            loss function
+        optimizer_params:
+            parameters for optimizer for Adam optimizer (api reference :py:class:`torch.optim.Adam`)
+        """
+        super().__init__()
+        self.patch_len = patch_len
+        self.num_layers = num_layers
+        self.feedforward_size = feedforward_size
+        self.hidden_size = hidden_size
+        self.nhead = nhead
+        self.stride = stride
+        self.loss = torch.nn.MSELoss() if loss is None else loss
```

Comment:
it's better to explicit `self.loss = loss` here because we pass this argument upstream